pkg/activator/throttler.go

```diff
@@ -97,23 +97,17 @@ func (t *Throttler) Try(rev RevisionID, function func()) error {
 	return nil
 }
 
-// This method updates Breaker's concurrency and requires external synchronization, `updateCapacity` requires `mux` to be held.
+// This method updates Breaker's concurrency.
 func (t *Throttler) updateCapacity(revision *v1alpha1.Revision, breaker *queue.Breaker, size int32) (err error) {
 	cc := int32(revision.Spec.ContainerConcurrency)
-	// The concurrency is unlimited, thus hand out as many tokens as we can in this breaker.
-	if cc == 0 {
-		cc = t.breakerParams.MaxConcurrency
-	}
-	delta := size*cc - breaker.Capacity()
-	// Do not update throttler's concurrency if the same number of endpoints is received
-	// or the number is negative.
-	if delta <= 0 {
-		return nil
-	}
-	if err := breaker.UpdateConcurrency(delta); err != nil {
-		return err
+
+	targetCapacity := cc * size
+	if targetCapacity == 0 {
+		// The concurrency is unlimited, thus hand out as many tokens as we can in this breaker.
+		targetCapacity = t.breakerParams.MaxConcurrency
 	}
-	return nil
+
+	return breaker.UpdateConcurrency(targetCapacity)
 }
 
 // getOrCreateBreaker retrieves existing breaker or creates a new one.
```

Comment:
Should we move the lock from `getOrCreateBreaker` to the parent/upper methods?
I just realise that we could have a potential race condition here - https://github.com/knative/serving/blob/4f885ac2deec829057b2289afa1587844ca2d49b/pkg/activator/throttler.go#L78-L79

Might be not directly related to this PR though, I could have a look at after this one gets in as well.

Reply:
As per our Slack discussion: This method is now designed to be called in parallel as much as you like.

However, the underlying mechanism to feed events into this of course needs to make sure that events arrive in the correct order. [Informers already do this today](https://github.com/kubernetes/client-go/issues/336) and the relevant function to be called once at a time thus is `UpdateEndpoints`. I added a comment there to state that assumption.